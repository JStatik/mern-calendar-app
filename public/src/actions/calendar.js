import { fetchEvents } from "../helpers/fetch";
import popUp from "../helpers/popup";
import prepareEvents from "../helpers/prepareEvents";
import types from "../types/types";

export const startLoadEvents = () => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( '/events', 'GET', token );
            const eventos = await peticion.json();

            const eventosPreparados = prepareEvents( eventos.eventos );

            dispatch( loadEvents( eventosPreparados ) );
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo obtener los eventos, intente nuevamente.' );
        }
    };
};

export const startDeleteEvent = () => {
    return async( dispatch, getState ) => {
        const { idEvent } = getState().calendar.activeEvent;
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( `/events/${ idEvent }`, 'DELETE', token );
            const evento = await peticion.json();

            if( evento.ok ) {
                dispatch( deleteEvent() );
            } else {
                dispatch( removeActiveEvent() );
                popUp( 'error', '#990000', 'Error', 'No tiene privilegios de eliminación sobre este evento' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo eliminar el evento, intente nuevamente.' );
        }
    };
};

export const startUpdateEvent = ( event ) => {
    return async( dispatch, getState ) => {
        const { idEvent } = getState().calendar.activeEvent;
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( `/events/${ idEvent }`, 'PUT', token, event );
            const evento = await peticion.json();

            if( evento.ok ) {
                event.user = evento.evento.user;
                event.idEvent = evento.evento.id;

                dispatch( updateEvent( event ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'No tiene privilegios de edición sobre este evento' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo actualizar el evento, intente nuevamente.' );
        }
    };
};

export const startAddNewEvent = ( event ) => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( '/events', 'POST', token, event );
            const evento = await peticion.json();

            if( evento.ok ) {
                event.user = evento.evento.user;
                event.idEvent = evento.evento.id;

                dispatch( addNewEvent( event ) );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo crear el evento, intente nuevamente.' );
        }
    };
};

const loadEvents = ( events ) => (
    {
        type: types.eventLoad,
        payload: events
    }
);

const deleteEvent = () => (
    {
        type: types.eventDelete
    }
);

const updateEvent = ( event ) => (
    {
        type: types.eventUpdate,
        payload: event
    }
);

const addNewEvent = ( event ) => (
    {
        type: types.eventAddNew,
        payload: event
    }
);

export const removeActiveEvent = () => (
    {
        type: types.eventRemoveActive
    }
);

export const setActiveEvent = ( event ) => (
    {
        type: types.eventSetActive,
        payload: event
    }
);

export const removeDatesSlot = () => (
    {
        type: types.eventRemoveDateSlot
    }
);

export const setDatesSlot = ( startSlot, endSlot ) => (
    {
        type: types.eventSetDateSlot,
        payload: {
            start: startSlot,
            end: endSlot
        }
    }
);

export const removeEventsLogout = () => (
    {
        type: types.eventRemove
    }
);
