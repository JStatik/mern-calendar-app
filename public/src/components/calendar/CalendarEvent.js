import React from 'react';

const CalendarEvent = ( event ) => {
    const { title } = event;
    const { user, notes } = event.event;

    return (
        <div>
            <strong>{ title }</strong><span> - { user.name }</span>
            <br/>
            <span><em>{ notes }</em></span>
        </div>
    )
};

export default CalendarEvent;
