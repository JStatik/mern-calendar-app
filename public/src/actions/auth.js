import types from '../types/types';
import { fetchLoginAndRegister, fetchRenew } from '../helpers/fetch';
import popUp from '../helpers/popup';
import { removeEventsLogout } from './calendar';

export const startChecking = () => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchRenew( '/auth/renew', token );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                dispatch( checkingFinish() );
                dispatch( startLogout() );
            }
        } catch( error ) {
            localStorage.removeItem( 'tkca' );
        }
    };
};

export const startRegister = ( name, email, password ) => {
    return async( dispatch ) => {
        const data = {
            name: name,
            email: email,
            password: password
        };

        try {
            const peticion = await fetchLoginAndRegister( '/auth/register', data );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'El email ingresado ya está en uso.' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'Error al registrarse, intente nuevamente.' );
        }
    };
};

export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        const data = {
            email: email,
            password: password
        };

        try {
            const peticion = await fetchLoginAndRegister( '/auth', data );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'Error al iniciar sesión, intente nuevamente.' );
        }
    };
};

export const startLogout = () => {
    return ( dispatch ) => {
        localStorage.removeItem( 'tkca' );

        dispatch( removeEventsLogout() );              
        dispatch( logout() );
    };
};

export const checkingFinish = () => (
    {
        type: types.authCheckingFinish
    }
);

const login = ( user ) => (
    {
        type: types.authLogin,
        payload: user
    }
);

const logout = () => (
    {
        type: types.authLogout
    }
);
