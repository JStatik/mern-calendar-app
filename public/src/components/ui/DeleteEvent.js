import React from 'react';
import { useDispatch } from 'react-redux';
import { startDeleteEvent } from '../../actions/calendar';

const DeleteEvent = () => {
    const dispatch = useDispatch();

    const handleClick = () => {
        dispatch( startDeleteEvent() );
    };

    return (
        <button className="btn btn-danger fab-danger" onClick={ handleClick }><i className="fa fa-trash"></i></button>
    )
};

export default DeleteEvent;
