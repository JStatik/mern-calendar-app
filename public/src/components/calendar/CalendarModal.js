import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Modal from 'react-modal';
import moment from 'moment';
import DateTimePicker from 'react-datetime-picker';
import { uiCloseModal } from '../../actions/ui';
import { removeActiveEvent, removeDatesSlot, startAddNewEvent, startUpdateEvent } from '../../actions/calendar';
import useForm from '../../hooks/useForm';
import { isFormValidEvents } from '../../helpers/isFormValid';

const customStyles = {
    content : {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate( -50%, -50% )'
    }
};

process.env.NODE_ENV !== 'test' && Modal.setAppElement( '#root' );

const now = moment().minutes( 0 ).seconds( 0 ).add( 1, 'hours' );
const nowPlus = now.clone().add( 1, 'hours' );

const initialFormValues = {
    start: now.toDate(),
    end: nowPlus.toDate(),
    title: '',
    notes: ''
};

const CalendarModal = () => {
    const dispatch = useDispatch();
    const { openModal } = useSelector( store => store.ui );
    const { activeEvent, slotDates } = useSelector( store => store.calendar );

    const [ dateStart, setDateStart ] = useState( now.toDate() );
    const [ dateEnd, setDateEnd ] = useState( nowPlus.toDate() );
    const [ msgError, setMsgError ] = useState( null );
    const [ datesValid, setDatesValid ] = useState( true );
    const [ titleValid, setTitleValid ] = useState( true );
   
    const [ formValues, setFormValues, handleInputChange, reset ] = useForm( initialFormValues );
    const { start, end, title, notes } = formValues;

    useEffect( () => {
        if( activeEvent ) {
            const { start, end, title, notes } = activeEvent;

            setFormValues( { start, end, title, notes } );      
            setDateStart( start );
            setDateEnd( end );
        } else if( slotDates ) {
            const { start, end } = slotDates;

            setFormValues( { ...initialFormValues, start: start, end: end } );
            setDateStart( start );
            setDateEnd( end );
        } else {
            setFormValues( initialFormValues );
            setDateStart( now.toDate() );
            setDateEnd( nowPlus.toDate() );
        }
    }, [ setFormValues, activeEvent, slotDates ] );

    const closeModal = () => {
        dispatch( removeActiveEvent() );
        dispatch( removeDatesSlot() );
        dispatch( uiCloseModal() );
        
        setDateStart( now.toDate() );
        setDateEnd( nowPlus.toDate() );

        setMsgError( null );
        setDatesValid( true );
        setTitleValid( true );
        reset();
    };

    const handleStartDateChange = ( event ) => {
        setDateStart( event );
        setFormValues( { ...formValues, start: event } );
    };

    const handleEndDateChange = ( event ) => {
        setDateEnd( event );
        setFormValues( { ...formValues, end: event } );
    };

    const handleSubmit = ( event ) => {
        event.preventDefault();

        const momentStart = moment( start );
        const momentEnd = moment( end );

        const { newMsgError, datesValid, titleValid, isValid } = isFormValidEvents( momentStart, momentEnd, title );

        setMsgError( newMsgError );
        setDatesValid( datesValid );
        setTitleValid( titleValid );

        if( isValid ) {
            if( activeEvent ) {
                dispatch( startUpdateEvent( formValues ) );
            } else {
                dispatch( startAddNewEvent( formValues ) );
            }

            closeModal();
        }
    };

    return (
        <Modal isOpen={ openModal } onRequestClose={ closeModal } closeTimeoutMS={ 200 } style={ customStyles } className="modal animate__animated animate__fadeIn" overlayClassName="modal-fondo" ariaHideApp={ process.env.NODE_ENV === 'test' ? false : true }>
            <h3 className="modal-title">{ activeEvent ? 'Editar evento' : 'Nuevo evento' }</h3>
            <hr/>
            <form className="container" autoComplete="off" onSubmit={ handleSubmit }>
                { msgError && <div className="alert-error">{ msgError }</div> }

                <div className="form-group">
                    <label>Fecha y hora de inicio</label>
                    <DateTimePicker className={ `form-control ${ !datesValid && 'is-invalid' }` } onChange={ handleStartDateChange } value={ dateStart } />
                </div>

                <div className="form-group">
                    <label>Fecha y hora de fin</label>
                    <DateTimePicker className={ `form-control ${ !datesValid && 'is-invalid' }` } onChange={ handleEndDateChange } value={ dateEnd } minDate={ dateStart } />
                </div>
                <hr/>
                <div className="form-group">
                    <label>Titulo y notas</label>
                    <input type="text" className={ `form-control ${ !titleValid && 'is-invalid' }` } placeholder="Título del evento" name="title" onChange={ handleInputChange } value={ title }/>
                    <small id="emailHelp" className="form-text text-muted">Una descripción corta</small>
                </div>

                <div className="form-group">
                    <textarea type="text" className="form-control" placeholder="Notas" rows="5" name="notes" onChange={ handleInputChange } value={ notes }></textarea>
                    <small id="emailHelp" className="form-text text-muted">Información adicional (Opcional)</small>
                </div>

                <button type="submit" className="btn btn-outline-primary btn-block">
                    <i className="far fa-save"></i>
                    <span> Guardar</span>
                </button>
            </form>
        </Modal>
    )
};

export default CalendarModal;
