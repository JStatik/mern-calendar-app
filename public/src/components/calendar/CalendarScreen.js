import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Calendar, momentLocalizer } from 'react-big-calendar';
import moment from 'moment';
import { uiOpenModal } from '../../actions/ui';
import { removeActiveEvent, setActiveEvent, setDatesSlot, startLoadEvents } from '../../actions/calendar';
import Navbar from '../ui/Navbar';
import CalendarEvent from './CalendarEvent';
import AddNewEvent from '../ui/AddNewEvent';
import DeleteEvent from '../ui/DeleteEvent';
import CalendarModal from './CalendarModal';
import messages from '../../helpers/calendar_messages_es';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import 'moment/locale/es';

moment.locale( 'es' );

const localizer = momentLocalizer( moment );

const CalendarScreen = () => {
    const dispatch = useDispatch();
    const { uid } = useSelector( store => store.auth );
    const { events, activeEvent } = useSelector( store => store.calendar );

    const [ viewCalendar, setViewCalendar ] = useState( localStorage.getItem( 'lastView' ) || 'month' );

    useEffect( () => {
        dispatch( startLoadEvents() );
    }, [ dispatch ] );

    const onViewCurrent = ( event ) => {
        setViewCalendar( event );
        localStorage.setItem( 'lastView', event );
    };

    const onSelect = ( event ) => {
        dispatch( setActiveEvent( event ) );
    };

    const onDoubleClick = () => {
        dispatch( uiOpenModal() );
    };

    const onSelectSlot = ( event ) => {
        dispatch( removeActiveEvent() );

        if( event.action === 'doubleClick' ) {
            let { start, end } = event;

            end = moment( end ).add( 1, 'hours' ).toDate();

            dispatch( setDatesSlot( start, end ) );
            dispatch( uiOpenModal() );
        }
    };

    const eventStyleGetter = ( event ) => {
        const style = {
            background: event.user.uid === uid ? '#2F2140' : '#7F69A5',
            color: '#ffffff',
            opacity: 0.85
        }

        return { style: style }
    };

    return (
        <div className="calendar-screen animate__animated animate__fadeIn">
            <Navbar />

            <Calendar 
                localizer={ localizer } 
                events={ events } 
                startAccessor="start" 
                endAccessor="end" 
                messages={ messages } 
                view={ viewCalendar } 
                components={ { event: CalendarEvent } } 
                eventPropGetter={ eventStyleGetter } 
                onView={ onViewCurrent } 
                onSelectEvent={ onSelect } 
                onDoubleClickEvent={ onDoubleClick } 
                onSelectSlot={ onSelectSlot } 
                selectable={ true } 
            />

            <AddNewEvent />
            { activeEvent && <DeleteEvent /> }

            <CalendarModal />
        </div>
    )
};

export default CalendarScreen;
