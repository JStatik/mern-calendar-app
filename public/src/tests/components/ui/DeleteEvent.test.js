import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import DeleteEvent from '../../../components/ui/DeleteEvent';
import { startDeleteEvent } from '../../../actions/calendar';
jest.mock( '../../../actions/calendar', () => (
    {
        startDeleteEvent: jest.fn()
    }
) );

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const store = mockStore( {} );
store.dispatch = jest.fn();

describe( 'Pruebas en DeleteEvent', () => {
    const wrapper = mount( 
        <Provider store={ store }>
            <DeleteEvent />
        </Provider>
    );

    test( 'Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    } );

    test( 'Debe ejecutar el dispatch de la accion startDeleteEvent()', () => {
        wrapper.find( 'button' ).simulate( 'click' );

        expect( store.dispatch ).toHaveBeenCalled();
        expect( startDeleteEvent ).toHaveBeenCalled();
    } );
} );
