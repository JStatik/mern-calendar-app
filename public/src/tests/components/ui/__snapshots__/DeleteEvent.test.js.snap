// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en DeleteEvent Debe mostrar el componente correctamente 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fa fa-trash"
  />
</button>
`;
